File: python/example1a.py

class ShoppingBasket:

    def __init__(self):
        self.items = {}

    def add(self, item, quantity):
        self.items[item.name] = (item, quantity)

    def get_quantity(self, item):
        if item.name in self.items.keys():
            item, quantity = self.items[item.name]
            return quantity
        return None

    def calculate_total(self):
        total = 0
        for (item, quantity) in self.items.values():
            total += item.price * quantity

        if total > 100:
            total = total * 0.95
        return total


class BasketItem:

    def __init__(self, name, price):
        self.name = name
        self.price = price

File: cpp/test-catch2/example1a.cpp
#include "catch2/catch.hpp"
#include "example1a.h"


TEST_CASE ("ShoppingBasket") {
    SECTION("total over 100 gives five percent discount") {
        auto basket = ShoppingBasket();
        auto itemA = BasketItem("A", 10);
        auto itemB = BasketItem("B", 25);
        auto itemC = BasketItem("C", 9.99);
        basket.add(itemA, 5);
        basket.add(itemB, 2);
        basket.add(itemC, 6);
        CHECK(basket.getQuantity(itemC) == 6);
        CHECK(basket.calculateTotal() == Approx(151.94).margin(0.01));
    }
}


